### PR TITLE
Forwarder affiliate fee usability improvements

### PR DESCRIPTION
**Pull request by abandeali1**
## Description

- Changes the affiliate fee to a flat amount rather than a percentage
- Removes the cap on affiliate fee size (it isn't possible to deterministically calculate what % of trade size the new flat affiliate fee would be)
- Splits the affiliate fee between multiple recipients

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/exchange-forwarder/contracts/src/MixinWeth.sol [hunk 1]
```diff
@@ -42,78 +41,77 @@ contract MixinWeth is
         }
     }
 
-    /// @dev Converts message call's ETH value into WETH.
-    function _convertEthToWeth()
+    /// @dev Transfers ETH denominated fees to all feeRecipient addresses
+    /// @param ethFeeAmounts Amounts of ETH, denominated in Wei, that are payed to corresponding feeRecipients.
```
Comment (dorothy-zbornak) [inline contracts/exchange-forwarder/contracts/src/MixinWeth.sol]:
```suggestion
    /// @param ethFeeAmounts Amounts of ETH, denominated in Wei, that are paid to corresponding feeRecipients.
```
:canada:

File: contracts/integrations/test/forwarder/forwarder_test.ts [hunk 2]
```diff
@@ -605,32 +649,39 @@ blockchainTests('Forwarder integration tests', env => {
...
-                forwarderFeePercentage,
-                revertError,
-            });
+            return expect(
+                forwarder
+                    .marketBuyOrdersWithEth([order], order.makerAssetAmount, [order.signature], forwarderFeeAmounts, [
+                        randomAddress(),
+                    ])
+                    .awaitTransactionSuccessAsync({ value }),
+            ).to.revertWith(revertError);
```
Comment (mzhu25) [inline contracts/integrations/test/forwarder/forwarder_test.ts]: can we pass the revert error to `marketBuyTestAsync` here? With `fractionalNumberOfOrdersToFill=0` and `ethValueAdjustment: -1`

File: contracts/integrations/test/forwarder/forwarder_test_factory.ts [hunk 3]
```diff
@@ -90,19 +82,18 @@ export class ForwarderTestFactory {
...
                 orders.map(signedOrder => signedOrder.signature),
-                feePercentage,
-                this._forwarderFeeRecipient.address,
+                forwarderFeeAmounts,
+                forwarderFeeRecipientAddresses,
             )
             .awaitTransactionSuccessAsync({
-                value: wethSpentAmount.plus(ethSpentOnForwarderFee).plus(ethValueAdjustment),
+                value: wethSpentAmount
+                    .plus(forwarderFeeAmounts.reduce((prev, curr) => prev.plus(curr), constants.ZERO_AMOUNT))
```
[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/integrations/test/forwarder/forwarder_test_factory.ts]: I ended up leaving this for now. `BigNumber.sum` doesn't handle no arguments well for some reason, and handling that is just about as verbose as using `reduce`.

[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/integrations/test/forwarder/forwarder_test_factory.ts]: maybe `BigNumber.sum(...[0, ...forwarderFeeAmounts])`

**Pull request merged**